packages/flutter/lib/src/material/input_decorator.dart

```diff
@@ -2325,9 +2326,11 @@ class _InputDecoratorState extends State<InputDecorator> with TickerProviderStat
               ),
               child: IconButtonTheme(
                 data: IconButtonThemeData(
-                style: IconButton.styleFrom(
-                  foregroundColor: _getPrefixIconColor(themeData, defaults),
-                  iconSize: iconSize,
+                  style: iconButtonTheme.style?.merge(
```

Comment:
Sorry for the late response here!

I think maybe there is actually a problem that is being exposed by the Google test failures. The MaterialStateProperty colors in the iconButtonTheme here are not having the [materialState of the InputDecorator](https://github.com/flutter/flutter/blob/29ed214e7ffbf9a0aff94d31a691ee12ac07c253/packages/flutter/lib/src/material/input_decorator.dart#L2139) applied to them. See the example below.

<details>

<summary>I would expect the icon to be purple in this example, but it's lime</summary>

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
        iconButtonTheme: IconButtonThemeData(
          style: ButtonStyle(
            foregroundColor: WidgetStateProperty.resolveWith((Set<WidgetState> states) {
              if (states.contains(WidgetState.error)) {
                return Colors.deepPurple;
              }
              return Colors.limeAccent;
            }),
          ),
        ),
      ),
      home: const MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatelessWidget {
  const MyHomePage({Key? key, required this.title}) : super(key: key);

  final String title;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(title),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            TextField(
              decoration: InputDecoration(
                errorText: 'error state',
                suffixIcon: IconButton(
                  onPressed: () {
                  },
                  icon: const Icon(Icons.phonelink_erase),
                ),
              ),
            ),
          ],
        ),
      ),
    );
  }
}
```

</details>

Reply:
Thanks so much for the details!

It turns out the inherited `IconButtonTheme.style.foreground` couldn't be overridden by `_getPrefixIconColor` & `_getPrefixIconColor` and resolve input decoration states.

I've updated both `_getPrefixIconColor` & `_getSuffixIconColor` to inherit the `IconButtonTheme.style.foreground` and resolve the state while placing in the order it can be superseded by `inputDecorationTheme` and `decoration.suffixIconColor`.

> I think maybe there is actually a problem that is being exposed by the Google test failures

Add two new tests to verify ``IconButtonTheme.style.foreground` color respects input decoration states  in the `input_decorator_test.dart`.